Zend/zend_execute.c

```diff
@@ -1532,13 +1532,15 @@ static zend_never_inline void zend_assign_to_string_offset(zval *str, zval *dim,
 	size_t string_len;
 	zend_long offset;
 	zend_string *s;
+	bool is_original_str = Z_REFCOUNTED_P(str) && Z_REFCOUNT_P(str) == 1;
+	zend_string *original_str = Z_STR_P(str);
 
 	/* separate string */
-	if (Z_REFCOUNTED_P(str) && Z_REFCOUNT_P(str) == 1) {
-		s = Z_STR_P(str);
+	if (is_original_str) {
+		s = original_str;
```

Comment:
I'm not even sure if this or the previous implementations were fully correct, since an error handler could theoretically replace the reference to a string in a reference group with a different value, possibly a non-string. 

(the string would have refcount 1, the reference would have refcount 2)

```php
<?php
function test(string $x) {
    $x = strtoupper($x);
    $y = [&$x];
    debug_zval_dump($y);
    $x[0] = 'x'; // if this was undefined, an error handler could be triggered and modify shared references (e.g. if $y were global)
}
test($argv[0]);
```

Reply:
I already committed the fix. See df16da36972c8e186cfe22433d5d0f278e6d7454. I forgot to decrement reference cuonter of the original string during separation.